src/nccl_ofi_mr.c

```diff
@@ -108,14 +108,13 @@ static inline void compute_page_address(uintptr_t addr,
 }
 
 void *nccl_ofi_mr_cache_lookup_entry(nccl_ofi_mr_cache_t *cache,
-				     void *addr,
-				     size_t size)
+									 const nccl_ofi_mr_input_t *in)
 {
 	uintptr_t page_addr;
 	size_t pages;
 
-	compute_page_address((uintptr_t)addr,
-			     size,
+	compute_page_address(nccl_ofi_mr_input_base(in),
```

Comment:
I think this is the wrong interface.  The lookup is always virtual address based.  There's no reason on a lookup function to change prototypes.  The lookup is still always a base and bound.

Reply:
The interface is still a base and a bound, it's just defined in terms of `_base` and `_len` functions against an opaque type, which stops us from needing to add two additional arguments to virtually every single function in the codebase and where those arguments are potentially invalid. 

by accepting that opaque type at the interface level, it prevents callers from generating base/len inconsistently depending on whether the underlying mr request from NCCL had dmabuf arguments or not.

Reply:
I don't think that's a win.

Reply:
No correct usage of this API ever involves querying the cache with a base address that doesn't have the correct size associated with it, or with the wrong base address entirely. Nobody should ever query the cache, then modify the base address or length and try it again. There's a single correct `{ addr, len }` key to query the cache with for any given MR and every other query pair is just a bug.

You're more or less directly arguing for leaking an abstraction. It doesn't matter that the internal API of the mr cache is `{ addr, len }`, what every consumer of this API wants is to pass in the single correct key for the specific MR that's in front of them, or some type that can derive that single correct key, and never anything else.

So this defines a type that encapsulates the arguments from NCCL ("the MR in front of them"), it constructs it immediately on the NCCL top level regmr call, it forwards it along through the MR codepath where we previously passed addr + len pairs separately, it defines a pair of functions that can generate the cache key from that encapsulation (`nccl_ofi_mr_input_base`/`nccl_ofi_mr_input_len`), and then it uses it. And it should have zero overhead.

The goal here isn't to optimize MR cache query submission, it's to prevent code duplication and consolidate dmabuf/iovec differences in the rest of the MR codepath. But it's a nice side effect that it also makes it difficult to query the cache incorrectly.